File: app/src/main/java/com/settlex/android/presentation/transactions/adapter/RecipientAdapter.java
package com.settlex.android.presentation.transactions.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.settlex.android.data.remote.profile.ProfileService;
import com.settlex.android.databinding.ItemRecipientBinding;
import com.settlex.android.presentation.transactions.model.RecipientUiModel;

public class RecipientAdapter extends ListAdapter<RecipientUiModel, RecipientAdapter.SuggestionsViewHolder> {
    private OnItemClickListener onItemClickListener;

    public interface OnItemClickListener {
        void onItemClick(RecipientUiModel model);
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.onItemClickListener = listener;
    }

    public RecipientAdapter() {
        super(DIFF_CALLBACK);
    }

    private static final DiffUtil.ItemCallback<RecipientUiModel> DIFF_CALLBACK = new DiffUtil.ItemCallback<>() {
        @Override
        public boolean areItemsTheSame(@NonNull RecipientUiModel oldItem, @NonNull RecipientUiModel newItem) {
            return oldItem.paymentId.equals(newItem.paymentId);
        }

        @Override
        public boolean areContentsTheSame(@NonNull RecipientUiModel oldItem, @NonNull RecipientUiModel newItem) {
            return oldItem.equals(newItem);
        }
    };

    @NonNull
    @Override
    public SuggestionsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemRecipientBinding binding = ItemRecipientBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new SuggestionsViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull SuggestionsViewHolder holder, int position) {
        holder.bind(getItem(position), onItemClickListener);
    }

    public static class SuggestionsViewHolder extends RecyclerView.ViewHolder {
        private final ItemRecipientBinding binding;

        public SuggestionsViewHolder(@NonNull ItemRecipientBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(RecipientUiModel recipient, OnItemClickListener listener) {
            ProfileService.loadProfilePic(recipient.photoUrl, binding.ivProfilePhoto);
            binding.tvFullName.setText(recipient.fullName);
            binding.tvPaymentId.setText(recipient.paymentId);
            binding.getRoot().setOnClickListener(v -> listener.onItemClick(recipient));
        }
    }
}

File: app/src/main/java/com/settlex/android/presentation/transactions/adapter/RecipientAdapter.kt
package com.settlex.android.presentation.transactions.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.settlex.android.data.remote.profile.ProfileService.loadProfilePhoto
import com.settlex.android.databinding.ItemRecipientBinding
import com.settlex.android.presentation.common.extensions.addAtPrefix
import com.settlex.android.presentation.transactions.adapter.RecipientAdapter.SuggestionsViewHolder
import com.settlex.android.presentation.transactions.model.RecipientUiModel

class RecipientAdapter(private val onItemClickListener: OnItemClickListener) :
    ListAdapter<RecipientUiModel, SuggestionsViewHolder>(DIFF_CALLBACK) {

    interface OnItemClickListener {
        fun onClick(selectedRecipient: RecipientUiModel)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SuggestionsViewHolder {
        val binding =
            ItemRecipientBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return SuggestionsViewHolder(binding)
    }

    override fun onBindViewHolder(holder: SuggestionsViewHolder, position: Int) {
        holder.bind(getItem(position), onItemClickListener)
    }

    class SuggestionsViewHolder(private val binding: ItemRecipientBinding) :
        RecyclerView.ViewHolder(binding.root) {

        fun bind(recipient: RecipientUiModel, listener: OnItemClickListener) = with(binding) {
            loadProfilePhoto(recipient.photoUrl, ivProfilePhoto)
            tvFullName.text = recipient.fullName
            tvPaymentId.text = recipient.paymentId.addAtPrefix()
            root.setOnClickListener { listener.onClick(recipient) }
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<RecipientUiModel>() {
            override fun areItemsTheSame(
                oldItem: RecipientUiModel,
                newItem: RecipientUiModel
            ): Boolean {
                return oldItem.paymentId == newItem.paymentId
            }

            override fun areContentsTheSame(
                oldItem: RecipientUiModel,
                newItem: RecipientUiModel
            ): Boolean {
                return oldItem == newItem
            }
        }
    }
}
